### PR TITLE
Init network

### PR DESCRIPTION
**Pull request by Vincinator**
Hi, 

This PR uses the crate [little-dude/netlink](https://github.com/little-dude/netlink) to initialize network devices

* Added network initialization for the libvirt-pid1 initialization flow
    * Initialize `lo` device
        * assign `::1/128`
        * assign `127.0.0.1/8`
    * Initialize `eth0` device
        * assign `fe80::2/64`
    * Print network info after initialization
* Functions used in above initialization flow are a starting point for other scenarios
* Possibility to create sriov devices for a given network interface

## How to test

As usual, copy paste steps from [hack/README.md](https://github.com/aurae-runtime/auraed/blob/main/hack/README.md):

```
make build-container
make kernel
make initramfs

# create `vm-br0` bridge on your machine:
make network

# run auraed in a VM as pid 1:
make virsh-start virsh-console virsh-stop

# exit VM console with Ctrl+]
```

On a second terminal you are now able to ping the auraed node via
```
# name of network interface "vm-br0" is set up by make network
ping fe80::2%vm-br0 
```

## Screenshot
2022-09-29_19-54-57_screenshot

### CODE REVIEW

File: src/init/network.rs [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    let num = cmp::min(limit, sriov_totalvfs);
+
+    fs::write(
+        format!("/sys/class/net/{}/device/sriov_numvfs", iface),
+        num.to_string(),
+    )
+    .expect("Unable to write file");
+}
+
+pub async fn init_iface(
```
Comment (krisnova) [inline src/init/network.rs]: function signatures should denote the IP stack (IPv4) or (IPv6)

```rust
pub async fn init_iface_ipv6();
```

File: src/init/network.rs [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+    }
+
+    setup_sriov(iface, num_sriov);
+}
+
+pub async fn init_lo_network(handle: rtnetlink::Handle) {
+    let localhost_ipv4: IpNetwork = "127.0.0.1/8".parse().unwrap();
+    let localhost_ipv6: IpNetwork = "::1/128".parse().unwrap();
+
+    if let Err(e) = add_address("lo", localhost_ipv4, handle.clone()).await {
```
Comment (krisnova) [inline src/init/network.rs]: move "lo" to constant at top of file as we will likely have multiple loopback names in the future

File: src/init/network.rs [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+        error!("{:?}", e);
+    }
+
+    if let Err(e) = add_address("lo", localhost_ipv6, handle.clone()).await {
+        error!("{:?}", e);
```
Comment (krisnova) [inline src/init/network.rs]: include IP stack in error message please 

```rust
        error!("IPv6 Error: {:?}", e); // or similar
```

File: src/init/network.rs [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+                                info!(
+                                    "\t ipv6: {}",
+                                    convert_ipv6_to_string(addr)
+                                        .unwrap_or_else(|_| {
+                                            "<error converting ip>".to_string()
+                                        })
+                                );
+                            }
+                        }
+                        None => {}
```
Comment (krisnova) [inline src/init/network.rs]: Should we log here?

File: src/lib.rs [hunk 5]
```diff
@@ -171,30 +176,43 @@ pub struct SystemRuntime {
...
-        // Implement & Discuss (TODO): init::init_network();
         // Show available network interfaces
-        // fileio::show_dir("/sys/class/net/", false);
+        //show_dir("/sys/class/net/", false);
+
+        trace!("configure network");
+        let (connection, handle, _) = new_connection().unwrap();
+        tokio::spawn(connection);
+
+        trace!("configure lo");
```
Comment (krisnova) [inline src/lib.rs]: use "looback" for logs as "lo" is subject to change in future PRs

Comment (krisnova) [inline src/lib.rs]: use "looback" for logs as "lo" is subject to change in future PRs

**Review by krisnova — CHANGES_REQUESTED** [review]
Please see my notes 👀

**Comment by Vincinator** [comment]
Implemented your feedback

I removed the `init_iface` functions completely now.
I think the initialization code of a given scenario can directly call the network functions.

File: src/init/network.rs (hunk 1 above)
Comment (Vincinator) [reply src/init/network.rs]: I removed the init_iface functions completely now.
I think the initialization code of a given scenario can directly call the network functions.

File: src/init/network.rs (hunk 4 above)
Comment (Vincinator) [reply src/init/network.rs]: yes!

File: src/lib.rs (hunk 5 above)
Comment (Vincinator) [reply src/lib.rs]: done

File: src/init/network.rs (hunk 3 above)
Comment (Vincinator) [reply src/init/network.rs]: good idea. Added ipstack in error message in result checking of `add_address_ipv6` and `add_address_ipv4` calls.

The `init_iface` function is removed

**Review by krisnova — APPROVED** [review]
LGTM -- let's keep moving forward

**Pull request merged**